File: ts/src/kernels.test.ts
import { describe, expect, it } from "vitest";

import {
  FIXED_SPLITS,
  matmulInvariant,
  splitsForBatch,
  sumFlat,
  sumSplit,
} from "./kernels";
import { Model, logitsAlone, logitsInBatch } from "./server";
import { flipCase } from "./flipCase";
import { Rng, vecEqual } from "./testUtil";

describe("sumFlat", () => {
  it("is zero for an empty vector", () => {
    expect(sumFlat([])).toBe(0.0);
  });

  it("returns the single element", () => {
    expect(sumFlat([42.5])).toBe(42.5);
  });

  it("sums strictly left-to-right", () => {
    expect(sumFlat([0.1, 0.2, 0.3, 0.4])).toBe(0.1 + 0.2 + 0.3 + 0.4);
  });
});

describe("sumSplit", () => {
  it("differs from sumFlat in floating point when grouping changes", () => {
    const vals = [1e16, 1.0, -1e16, 1.0];
    expect(sumSplit(vals, 2)).not.toBe(sumFlat(vals));
  });

  it("equals sumFlat with a single split", () => {
    const vals = [0.1, 0.2, 0.3, 0.4];
    expect(sumSplit(vals, 1)).toBe(sumFlat(vals));
  });

  it("equals sumFlat for zero or negative splits", () => {
    const vals = [0.1, 0.2, 0.3];
    expect(sumSplit(vals, 0)).toBe(sumFlat(vals));
    expect(sumSplit(vals, -3)).toBe(sumFlat(vals));
  });

  it("is zero for an empty vector", () => {
    expect(sumSplit([], 4)).toBe(0.0);
  });

  it("handles more splits than elements", () => {
    expect(sumSplit([1.0, 2.0, 3.0], 8)).toBe(6.0);
  });
});

describe("splitsForBatch", () => {
  it.each([
    [0, 8],
    [1, 8],
    [2, 4],
    [4, 2],
    [8, 1],
    [9, 1],
    [16, 1],
  ])("maps batch %i to %i splits", (batch, expected) => {
    expect(splitsForBatch(batch)).toBe(expected);
  });

  it("varies with batch size", () => {
    expect(splitsForBatch(1)).not.toBe(splitsForBatch(4));
  });
});

describe("constants", () => {
  it("fixes the invariant split count at four", () => {
    expect(FIXED_SPLITS).toBe(4);
  });
});

describe("matmul", () => {
  it("computes the right products and shape", () => {
    const w = [
      [1.0, 2.0],
      [3.0, 4.0],
      [5.0, 6.0],
    ];
    const out = matmulInvariant(
      [
        [1.0, 1.0],
        [2.0, 2.0],
      ],
      w,
    );
    expect(out.length).toBe(2);
    expect(out[0]).toEqual([3.0, 7.0, 11.0]);
  });
});

describe("invariant kernel", () => {
  it("gives an identical row alone and batched over many batches", () => {
    const rng = new Rng(1);
    const model = new Model(rng.mixedRows(5, 16));
    for (let trial = 0; trial < 200; trial++) {
      const x = rng.mixedRow(16);
      const others = rng.mixedRows(trial % 7, 16);
      const alone = logitsAlone(model, x, true);
      const batched = logitsInBatch(model, x, others, true);
      expect(vecEqual(alone, batched)).toBe(true);
    }
  });
});

describe("variant kernel", () => {
  it("changes a row with the batch for the flip case", () => {
    const { model, x, others } = flipCase();
    expect(vecEqual(logitsAlone(model, x, false), logitsInBatch(model, x, others, false))).toBe(
      false,
    );
  });
});

File: go/kernels_test.go
package batchinvariant

import "testing"

func TestSumFlatEmptyIsZero(t *testing.T) {
	if got := SumFlat(Vec{}); got != 0.0 {
		t.Fatalf("SumFlat(empty) = %v, want 0", got)
	}
}

func TestSumFlatSingleElement(t *testing.T) {
	if got := SumFlat(Vec{42.5}); got != 42.5 {
		t.Fatalf("SumFlat([42.5]) = %v, want 42.5", got)
	}
}

func TestSumFlatLeftToRight(t *testing.T) {
	if got := SumFlat(Vec{0.1, 0.2, 0.3, 0.4}); got != ((0.1+0.2)+0.3)+0.4 {
		t.Fatalf("SumFlat is not left-to-right: %v", got)
	}
}

func TestSplitEqualsFlatInExactMathButNotInFloat(t *testing.T) {
	vals := Vec{1e16, 1.0, -1e16, 1.0}
	if SumSplit(vals, 2) == SumFlat(vals) {
		t.Fatalf("expected different grouping to change the float sum")
	}
}

func TestSingleSplitIsFlat(t *testing.T) {
	vals := Vec{0.1, 0.2, 0.3, 0.4}
	if SumSplit(vals, 1) != SumFlat(vals) {
		t.Fatalf("SumSplit(_, 1) must equal SumFlat")
	}
}

func TestZeroOrNegativeSplitsIsFlat(t *testing.T) {
	vals := Vec{0.1, 0.2, 0.3}
	if SumSplit(vals, 0) != SumFlat(vals) {
		t.Fatalf("SumSplit(_, 0) must equal SumFlat")
	}
	if SumSplit(vals, -3) != SumFlat(vals) {
		t.Fatalf("SumSplit(_, negative) must equal SumFlat")
	}
}

func TestSumSplitEmptyIsZero(t *testing.T) {
	if got := SumSplit(Vec{}, 4); got != 0.0 {
		t.Fatalf("SumSplit(empty, 4) = %v, want 0", got)
	}
}

func TestSumSplitMoreSplitsThanElements(t *testing.T) {
	vals := Vec{1.0, 2.0, 3.0}
	// size = ceil(3/8) = 1, so three singleton partials plus empties skipped.
	if got := SumSplit(vals, 8); got != 6.0 {
		t.Fatalf("SumSplit([1,2,3], 8) = %v, want 6", got)
	}
}

func TestSplitsForBatch(t *testing.T) {
	cases := map[int]int{0: 8, 1: 8, 2: 4, 4: 2, 8: 1, 9: 1, 16: 1}
	for batch, want := range cases {
		if got := SplitsForBatch(batch); got != want {
			t.Fatalf("SplitsForBatch(%d) = %d, want %d", batch, got, want)
		}
	}
}

func TestSplitsForBatchVariesWithBatchSize(t *testing.T) {
	if SplitsForBatch(1) == SplitsForBatch(4) {
		t.Fatalf("split count must vary with batch size")
	}
}

func TestMatmulShapes(t *testing.T) {
	w := Mat{{1, 2}, {3, 4}, {5, 6}}
	out := MatmulInvariant(Mat{{1, 1}, {2, 2}}, w)
	if len(out) != 2 || len(out[0]) != 3 {
		t.Fatalf("unexpected shape %dx%d", len(out), len(out[0]))
	}
	// Row 0 = [1*1+1*2, 1*3+1*4, 1*5+1*6] = [3, 7, 11].
	if out[0][0] != 3 || out[0][1] != 7 || out[0][2] != 11 {
		t.Fatalf("wrong products: %v", out[0])
	}
}

func TestInvariantRowIdenticalAloneAndBatched(t *testing.T) {
	r := newLCG(1)
	model := Model{W: r.mixedRows(5, 16)}
	for trial := 0; trial < 200; trial++ {
		x := r.mixedRow(16)
		others := r.mixedRows(trial%7, 16)
		alone := LogitsAlone(model, x, true)
		batched := LogitsInBatch(model, x, others, true)
		if !vecEqual(alone, batched) {
			t.Fatalf("invariant kernel differed on trial %d", trial)
		}
	}
}

func TestInvariantPositionInBatchDoesNotMatter(t *testing.T) {
	r := newLCG(2)
	w := r.mixedRows(4, 16)
	x := r.mixedRow(16)
	neighbours := r.mixedRows(2, 16)
	first := MatmulInvariant(append(Mat{x}, neighbours...), w)[0]
	alone := MatmulInvariant(Mat{x}, w)[0]
	if !vecEqual(first, alone) {
		t.Fatalf("invariant row depended on batch position")
	}
}

func TestVariantRowChangesWithBatch(t *testing.T) {
	// The constructed flip case is guaranteed to differ under the variant kernel.
	model, x, others := FlipCase()
	if vecEqual(LogitsAlone(model, x, false), LogitsInBatch(model, x, others, false)) {
		t.Fatalf("variant kernel should be batch-dependent")
	}
}
